Create enforcement.md - Encryption Enforcement topic

Andriciuc opened the pull request:
Writing a general topic regarding Encrpytion Enforcement from pg_tde perspective, requires heavy review.

Andriciuc committed:
initial encryption enforcement doc

populated with encryption enforcement

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-create-enforcement

Andriciuc committed:
Update enforcement.md

updated with future steps to enforce encrypt

Andriciuc committed:
Merge branch 'docs-create-enforcement' of https://github.com/percona/postgres into docs-create-enforcement

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-create-enforcement

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-create-enforcement

Andriciuc committed:
Update enforcement.md

updated with pg_tde enforce encryption option

Andriciuc committed:
Merge branch 'docs-create-enforcement' of https://github.com/percona/postgres into docs-create-enforcement

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-create-enforcement

dutow commented on contrib/pg_tde/documentation/docs/how-to/enforcement.md:
```diff
@@ -0,0 +1,15 @@
+
+The [`pg_tde.enforce_encryption`](../variables.md/#pg_tdeenforce_encryption) option automatically encrypts new tables as they are created. This removes the risk of sensitive information being accidentally exposed because it was not encrypted.
+
+### 2. Use encrypted tablespaces exclusively
```
What are encrypted tablespaces? That's a mysql concept, we don't have anything like that in postgres...

Andriciuc replied:
Removed!

dutow commented on contrib/pg_tde/documentation/docs/how-to/enforcement.md:
```diff
@@ -0,0 +1,15 @@
+
+### 1. Use pg_tde.enforce_encryption
+
+The [`pg_tde.enforce_encryption`](../variables.md/#pg_tdeenforce_encryption) option automatically encrypts new tables as they are created. This removes the risk of sensitive information being accidentally exposed because it was not encrypted.
```
It doesn't encrypt them automatically, it just prevents non encrypted tables to be created. If a user tries to create a "normal" non encrypted table, it returns an error instead.

Andriciuc replied:
Updated!

Andriciuc committed:
Update enforcement.md

removed table space encryption

Andriciuc committed:
Update enforcement.md

updated the enforce_encryption description

Andriciuc committed:
Update enforcement.md

added a note that clarifies unauthorized user access and how superusers can still bypass the forced encryption table creation

dutow commented on contrib/pg_tde/documentation/docs/how-to/enforcement.md:
```diff
@@ -0,0 +1,14 @@
+
+Use the following techniques to enforce the secure use of `pg_tde`.
+
+### 1. Use pg_tde.enforce_encryption
```
Now we have a first point without a second?

Andriciuc replied:
As said above, there will be future examples that I will add, thank you!

dutow commented on contrib/pg_tde/documentation/docs/how-to/enforcement.md:
```diff
@@ -0,0 +1,14 @@
+The [`pg_tde.enforce_encryption`](../variables.md/#pg_tdeenforce_encryption) option prevents non encrypted tables to be created.
+
+!!! note
+    This option prevents unauthorized users from creating unencrypted tables or indexes, ensuring encryption compliance. Superusers retain the privilege to create unencrypted tables as necessary.
```
No, superusers can override the variable at a session level, they also can't create unencrypted tables if the variable is enabled.

Andriciuc replied:
Oh! Thank you for the clarification, I thought superusers could override no matter the variable setting, fixed:

Superusers can override the `pg_tde.enforce_encryption` variable at the session level, but they still cannot create unencrypted tables while it is enabled.

dutow commented on contrib/pg_tde/documentation/docs/how-to/enforcement.md:
```diff
@@ -0,0 +1,14 @@
+
+## Enforcing encryption usage
+
+Use the following techniques to enforce the secure use of `pg_tde`.
```
From use the following techniques, I would expect some examples or use cases to follow - but we have neither here, just a repeated description.

Andriciuc replied:
I will start adding some use cases as per the above comment!

dutow reviewed (commented):
Now my generic problem with this PR is that this new page doesn't provide any additional information compared to the GUC page. In fact, it provides less, as the GUC page at least explains to an extent what are the different allowed scopes of the variable.

Jan asked for scenarios to be documented, I guess he meant things like:

* Specific example how to enforce encryption to the entire server
* Specific example how to enforce encryption to only a single database with `ALTER DATABASE SET`
* Example that superusers can override the variable at session level, so that they can create non encrypted tables in that session (the usecase I described on slack)
* Maybe also an example of enforcing encryption to a specific user with `ALTER USER`

Which is basically adding actual examples to the "Use the following techniques..." section

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-create-enforcement

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-create-enforcement

Andriciuc committed:
Update enforcement.md

small update to the note describing superuser actions properly

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-create-enforcement

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-create-enforcement

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-create-enforcement

Andriciuc committed:
Merge branch 'release-17.5.2' into docs-create-enforcement

Andriciuc committed:
Merge branch 'release-17.5.2' into docs-create-enforcement

Andriciuc committed:
Updated enforcement with examples and variables with link

Added a couple of simple examples to encryption enforcement for user, database and global enforcement using enforce_encryption

Added What does enforcement do (for new users)

Added that you can override enforcement with details possible by superusers only

Updated variables with minor linting and added a link to encryption enforcement for pg_tde.enforce_encryption

Andriciuc commented:
Added a couple of simple examples to encryption enforcement for user, database and global enforcement using enforce_encryption

Added What does enforcement do (for new users)

Added that you can override enforcement with details possible by superusers only

Updated variables with minor linting and added a link to encryption enforcement for pg_tde.enforce_encryption

Andriciuc committed:
Merge branch 'release-17.5.2' into docs-create-enforcement

Andriciuc committed:
Merge branch 'release-17.5.2' into docs-create-enforcement

dutow commented on contrib/pg_tde/documentation/docs/how-to/enforcement.md:
````diff
@@ -0,0 +1,62 @@
+pg_tde.enforce_encryption = on
+```
+
+This ensures that no user, including superusers, can create unencrypted tables unless they explicitly override the variable in a session (see below).
````
It's not clear to me that only superusers can override the variable, and because of that, it is misleading. Maybe instead:

> This ensures that no user, including superusers, can create unencrypted tables.
> Superusers can explicitly override the variable in a session (see below).

dutow replied:
Also it's never mentioned that only superusers can change this variable.

Andriciuc replied:
To fix this, I added a note to mention that **only** superusers can change this variable, I added a link to the override topic so the user can jump straight to it.

dutow commented on contrib/pg_tde/documentation/docs/how-to/enforcement.md:
```diff
@@ -0,0 +1,62 @@
+
+### Override enforcement for trusted sessions
+
+Superusers (such as DBAs) can override the variable at the session level:
```
`such as DBAs` - not needed / not accurate. Superusers are superuser accounts, regardless who uses them.

Andriciuc replied:
True! Removed DBAs mention, thank you!

Andriciuc committed:
removed dbas mention + updated 1. with note and link

removed dbas mention and updated 1. with a note regarding only superusers being able to change pg_tde.enforce_encryption

I also added a link to the override trusted sessions topic
